src/sonic-dhcp-server/dhcp_server/common/dhcp_db_monitor.py

```diff
@@ -1,131 +1,414 @@
-
-    def _do_check(self, check_param):
-        if "enabled_dhcp_interfaces" not in check_param:
-            syslog.syslog(syslog.LOG_ERR, "Cannot get enabled_dhcp_interfaces")
-            return (True, True, True)
-        enabled_dhcp_interfaces = check_param["enabled_dhcp_interfaces"]
-        return (self._check_dhcp_server_update(enabled_dhcp_interfaces),
-                self._check_vlan_update(enabled_dhcp_interfaces),
-                self._check_vlan_intf_update(enabled_dhcp_interfaces))
-
-    def _check_dhcp_server_update(self, enabled_dhcp_interfaces):
-        """
-        Check dhcp_server_ipv4 table
+            return (False, False, False)
+        return (self.checker_dict[DHCP_SERVER_IPV4].check_update_event(db_snapshot),
+                self.checker_dict[VLAN].check_update_event(db_snapshot),
+                self.checker_dict[VLAN_INTERFACE].check_update_event(db_snapshot))
+
+
+class DhcpServdDbMonitor(object):
+    checker_dict = {}
+    subscribed_table = set()
+
+    def __init__(self, db_connector, select_timeout=DEFAULT_SELECT_TIMEOUT):
+        self.db_connector = db_connector
+        self.sel = swsscommon.Select()
+        self.select_timeout = select_timeout
+        all_checker_classes = ConfigDbEventChecker.get_all_subclasses()
+        checker_list = [checker(self.sel) for checker in all_checker_classes
+                        if checker != DhcpServerTableIntfEnablementEventChecker]
+        for checker in checker_list:
+            self.checker_dict[checker.table_name] = checker
+
+    def unsubscribe_tables(self, unsubscribe_tables):
+        """
+        Unsubscribe monitor table change of tables
         Args:
-            enabled_dhcp_interfaces: DHCP interface that enabled dhcp_server
-        Returns:
-            Whether need to refresh
+            unsubscribe_tables: set contains name of tables need to be unsubscribed
         """
-        need_refresh = False
-        while self.subscribe_dhcp_server_table.hasData():
-            key, op, entry = self.subscribe_dhcp_server_table.pop()
-            if op == "SET":
-                for field, value in entry:
-                    if field != "state":
-                        continue
-                    # Only if new state is not consistent with old state, we need to refresh
-                    if key in enabled_dhcp_interfaces and value == "disabled":
-                        need_refresh = True
-                    elif key not in enabled_dhcp_interfaces and value == "enabled":
-                        need_refresh = True
-            # For del operation, we can skip disabled change
-            if op == "DEL":
-                if key in enabled_dhcp_interfaces:
-                    need_refresh = True
-        return need_refresh
+        for table in unsubscribe_tables:
+            if table not in self.checker_dict:
+                syslog.syslog(syslog.LOG_ERR, "Cannot find checker for {} in checker_dict".format(table))
```

Comment:
When this happen?

Reply:
Parameter is passed from dhcpservd. dhcpservd will dynamically subscribe/unsubscribe tables